libs/server-sdk/src/evaluation/bucketing.cpp

```diff
@@ -0,0 +1,130 @@
+#include "bucketing.hpp"
+
+#include <launchdarkly/detail/c_binding_helpers.hpp>
+#include <launchdarkly/encoding/base_16.hpp>
+#include <launchdarkly/encoding/sha_1.hpp>
+
+#include <sstream>
+#include <string_view>
+
+namespace launchdarkly::server_side::evaluation {
+
+double const kBucketScale = static_cast<double>(0x0FFFFFFFFFFFFFFF);
+
+AttributeReference const& Key();
+
+std::optional<ContextHashValue> ContextHash(Value const& value,
+                                            BucketPrefix prefix);
+
+std::optional<std::string> BucketValue(Value const& value);
+
+bool IsIntegral(double f);
+
+BucketPrefix::BucketPrefix(Seed seed) : prefix_(seed) {}
+
+BucketPrefix::BucketPrefix(std::string key, std::string salt)
+    : prefix_(KeyAndSalt{key, salt}) {}
+
+std::ostream& operator<<(std::ostream& os, BucketPrefix const& prefix) {
+    std::visit(
+        [&](auto&& arg) {
+            using T = std::decay_t<decltype(arg)>;
+            if constexpr (std::is_same_v<T, BucketPrefix::KeyAndSalt>) {
+                os << arg.key << "." << arg.salt;
+            } else if constexpr (std::is_same_v<T, BucketPrefix::Seed>) {
+                os << arg;
+            }
+        },
+        prefix.prefix_);
+    return os;
+}
+
+tl::expected<std::pair<ContextHashValue, RolloutKindLookup>, Error> Bucket(
+    Context const& context,
+    AttributeReference const& by_attr,
```

Comment:
I assume this is handled elsewhere, maybe in parsing, but when bucketBy is omitted entirely it is also "key".

Reply:
Yep, it's in the parser.